### PR TITLE
M diagnosis

### PR DESCRIPTION
**Pull request by macartan**
Reworking of declare_design and related functions

* separates simulations and diagnosis/bootstrapping into two steps: `simulate_design` and `diagnose_design`
* Note before both the simulations and diagnosis were done in chunks when there were multiple designs; now the simulations is done in chunks but the diagnosis is done in one go with the design_ID added to the `group_by` list
* `diagnose_design` can take designs as before but can also take a `simulation_df`
* `diagnosands_df` now have three attributes:  sims,  bootstrap, and number of diagnosands
*  Parameter attributes, e.g. generated by expand_design,  get passed to later dataframes by `diagnose_design`
* `diagnose_design` accepts additional grouping variables, to be used eg for diagnosis within fans or quick conditional diagnoses.
* `expand_design` can take a stub for naming designs
*  The print summary now reports sims and bootstrap; with sims as a vector when it is a vector 
* `reshape_design` combines Alex's code and mine to take advantage of information in attributes 
* `reshape_design` changes diagnosand labels to make them prettier
* `reshape_design` lets you choose diagnosands to be returned, by name
* `reshape_design` no longer depends on `se(` but uses attributes of the diagnosand_df to figure out what's what
* `diagnose_design` now prints the reshaped summary and information on sims and bootstraps
* `diagnose_design` now exports bootstraps, if these are calculated
*  warning if designs are sent of length not equal to sims length (when this is a vector)
*  warning when simulations_df  is longer than expected (eg because of repeated labels)
* documentation, commenting and related

### CODE REVIEW

File: R/diagnose_design.R [hunk 1]
```diff
@@ -41,41 +34,36 @@
...
+           add_parameters = FALSE) {
+    if (min(sims) < 1)
+      stop("Sims should be >= 1")
 
     designs <- list(...)
 
     inferred_names <- paste(substitute(list(...)))[-1]
 
-    ## three cases:
+    ## Two cases:
```
Comment (nfultz) [inline R/diagnose_design.R]: inferred names line should be replaced with rlang.

File: R/diagnose_design.R [hunk 2]
```diff
@@ -41,41 +34,36 @@
...
     ## 1. send one or more design objects created by declare_design
-    ## 2. send a single list of design objects created by expand_design
-    ## 3. do not allow sending more than one object if any of them aren't design objects.
-    if (length(designs) == 1 && is.list(designs[[1]]) && !"design" %in% class(designs[[1]]) ) {
-      ## this unpacks designs if a list of designs was sent as a single list object, i.e.
-      ##   as created by expand_design
+    ## 2. send a single list of design objects e.g. created by expand_design
+    ## Approach: unpack designs if a list of designs was sent as a single list object
+    if (length(designs) == 1 &&
+        is.list(designs[[1]]) && !"design" %in% class(designs[[1]])) {
```
Comment (nfultz) [inline R/diagnose_design.R]: use inherits

File: R/diagnose_design.R [hunk 3]
```diff
@@ -84,83 +72,135 @@ diagnose_design <- function(..., diagnosands = default_diagnosands,
...
+    # Ensure designs are length of sims if vector of sims is provided
+    if (!is.data.frame(sims))
+      if (length(sims) > 1) {
+        sim_length_check <- lapply(designs, length) != length(sims)
+        if (any(sim_length_check))
+          warning(paste0(
+            "sims is of different length to designs for designs: ",
+            paste0(names(designs[sim_length_check]), collapse = ", ")
+          ))
+      }
```
Comment (nfultz) [inline R/diagnose_design.R]: Pull these two checks in to helper functions

File: R/diagnose_design.R [hunk 4]
```diff
@@ -84,83 +72,135 @@ diagnose_design <- function(..., diagnosands = default_diagnosands,
...
+
+    simulations_list <-
+      Map(cbind,  design_ID = names(simulations_list), simulations_list)
+
+    # Create columns for non-existing attributes to allow merging
+    col_list <-
+      unique(unlist(as.vector(sapply(
+        simulations_list, names
+      ))))
+    simulations_list <- lapply(simulations_list, function(x) {
```
Comment (nfultz) [inline R/diagnose_design.R]: use rbind_disjoint instead.

File: R/diagnose_design.R [hunk 5]
```diff
@@ -84,83 +72,135 @@ diagnose_design <- function(..., diagnosands = default_diagnosands,
...
     }
 
-    simulations_df <- do.call(rbind, simulations_list)
-    diagnosands_df <- do.call(rbind, diagnosands_list)
+    # Add attributes
+    attr(simulations_df, "sims") <- sims
 
-    rownames(simulations_df) <- NULL
-    rownames(diagnosands_df) <- NULL
+    structure(simulations_df)
```
Comment (nfultz) [inline R/diagnose_design.R]: set sims = sims here instead of above.

File: R/diagnose_design.R [hunk 6]
```diff
@@ -84,83 +72,135 @@ diagnose_design <- function(..., diagnosands = default_diagnosands,
...
       results_list <- future_lapply(seq_len(sims),
-                                    function(i) run_design(design),
+                                    function(i)
+                                      run_design(design),
                                     future.seed = NA, future.globals = "design")
     } else {
       sims <- check_sims(design, sims)
       results_list <- fan_out(design, sims)
-      fan_id <- setNames(rev(do.call(expand.grid, lapply(rev(sims$n), seq))), paste0("fan_", seq_len(nrow(sims))))
+      fan_id <-
```
Comment (nfultz) [inline R/diagnose_design.R]: "variable arrow newline" is bad style, suggest

```
fan_id <- setNames(
```

instead

File: R/diagnose_design.R [hunk 7]
```diff
@@ -84,83 +72,135 @@ diagnose_design <- function(..., diagnosands = default_diagnosands,
...
+        cbind(sim_ID = rep(seq_along(subresult), vapply(subresult, nrow, 0L)), df)
     }
 
-
     estimates_df <- results2x(results_list, "estimates_df")
     estimands_df <- results2x(results_list, "estimands_df")
-    if (nrow(estimates_df) == 0 & nrow(estimands_df) == 0) {
-      stop("No estimates or estimands were declared, so diagnose_design cannot calculate diagnosands.", call. = FALSE)
-    }
+
```
Comment (nfultz) [inline R/diagnose_design.R]: Why no more error?

File: R/diagnose_design.R [hunk 8]
```diff
@@ -177,104 +217,301 @@ diagnose_design_single_design <- function(design, diagnosands, sims, bootstrap)
...
+#' \dontrun{
+#' simulations <- simulate_design(designs, sims = 2)
+#' diagnosis   <- diagnose_design(simulations_df = simulations)
+#'
+#' }
+
+#' @importFrom stats setNames
+#' @importFrom utils head
+#' @export
+diagnose_design <-
```
Comment (nfultz) [inline R/diagnose_design.R]: If we want this function to work with either lists of designs or simuation dfs, it should probably be generic instead. Split into diagnose_design.data.frame, which has all logic from the simulation forward, and diagnose_design.default, which creates simulations from designs/lists of designs, then passes that to diagnose_design.data.frame.

File: R/diagnose_design.R [hunk 9]
```diff
@@ -177,104 +217,301 @@ diagnose_design_single_design <- function(design, diagnosands, sims, bootstrap)
...
+      group_by_list <- simulations_df[, group_by_set, drop = FALSE]
 
       labels_df <- split(group_by_list, group_by_list, drop = TRUE)
-      labels_df <- lapply(labels_df, head, n=1)
-
-      diagnosands_df <- split(simulations_df, group_by_list, drop=TRUE)
-      diagnosands_df <- lapply(diagnosands_df, FUN = function(x){ dg <- diagnosands(x); setNames(dg[[2]], dg[[1]]) })
+      labels_df <- lapply(labels_df, head, n = 1)
+
+      diagnosands_df <-
```
Comment (nfultz) [inline R/diagnose_design.R]: this should be one line

File: R/diagnose_design.R [hunk 10]
```diff
@@ -177,104 +217,301 @@ diagnose_design_single_design <- function(design, diagnosands, sims, bootstrap)
...
+                                        add_parameters = add_parameters)
 
     group_by_set <- attr(diagnosands, "group_by") %||%
-      colnames(simulations_df) %i% c("estimand_label", "estimator_label", "coefficient")
+      colnames(simulations_df) %i% grouping_variables
 
+    ## Get the list of parameter attributes in design list since these are to be preserved in dataframe
+    if (add_parameters) {
+      group_by_set <- c(group_by_set, attr(simulations_df, "parameters"))
+    }
```
Comment (nfultz) [inline R/diagnose_design.R]: warn if add-parameters is  true but simulationsdf doesn't have the attribute.

File: R/diagnose_design.R [hunk 11]
```diff
@@ -177,104 +217,301 @@ diagnose_design_single_design <- function(design, diagnosands, sims, bootstrap)
...
+          FUN = function(x) {
+            dg <- diagnosands(x)
+            setNames(dg[[2]], dg[[1]])
+          }
+        )
 
-      diagnosands_df <- as.data.frame(t(mapply(c, labels_df, diagnosands_df)), stringsAsFactors=FALSE) # c appropriate and fast here bc labels_df must be a list (drop=FALSE above)
+      diagnosands_df <-
+        as.data.frame(t(mapply(c, labels_df, diagnosands_df)), stringsAsFactors =
+                        FALSE) # c appropriate and fast here bc labels_df must be a list (drop=FALSE above)
```
Comment (nfultz) [inline R/diagnose_design.R]: dangling false

File: R/diagnose_design.R [hunk 12]
```diff
@@ -177,104 +217,301 @@ diagnose_design_single_design <- function(design, diagnosands, sims, bootstrap)
...
       diagnosands_df[] <- lapply(diagnosands_df, unlist)
 
-      # diagnosands_df <- merge(labels_df, diagnosands_df, by = "row.names", all = TRUE, sort = FALSE)
-      # diagnosands_df$Row.names <- NULL
       return(diagnosands_df)
     }
 
-    diagnosands_df <- calculate_diagnosands(simulations_df, diagnosands)
+    diagnosands_df <-
```
Comment (nfultz) [inline R/diagnose_design.R]: one line

File: R/diagnosis_helper_functions.R [hunk 13]
```diff
@@ -73,15 +75,118 @@ summary.diagnosis <- function(object, ...) {
...
+      rbind_disjoint(list(diagnosands_only_df, se_only_df), infill = "")
+
+    # Reorder rows
+    nrows <- nrow(diagnosands_only_df)
+    return_df <-
+      return_df[rep(1:nrows, each = 2) + rep(c(0, nrows), nrows),]
+
+    # Select columns
+    if (!is.logical(selection))
+      if (any(!(selection %in% diagnosands_df_names)))
```
Comment (nfultz) [inline R/diagnosis_helper_functions.R]: demorgans law => !all()

File: R/diagnosis_helper_functions.R [hunk 14]
```diff
@@ -73,15 +75,118 @@ summary.diagnosis <- function(object, ...) {
...
+    return_df <-
+      rbind_disjoint(list(diagnosands_only_df, se_only_df), infill = "")
+
+    # Reorder rows
+    nrows <- nrow(diagnosands_only_df)
+    return_df <-
+      return_df[rep(1:nrows, each = 2) + rep(c(0, nrows), nrows),]
+
+    # Select columns
+    if (!is.logical(selection))
```
Comment (nfultz) [inline R/diagnosis_helper_functions.R]: prefer is.character over !is.logical, safer.

File: R/diagnosis_helper_functions.R [hunk 15]
```diff
@@ -73,15 +75,118 @@ summary.diagnosis <- function(object, ...) {
...
+        ))
+
+    return_df <- return_df[, selection]
+    return(return_df)
+
+  }
+
+# Get Names of Parameters from a list of designs
+get_parnames <- function(designs, sort_names = TRUE) {
+  out <- lapply(designs, function(j) {
```
Comment (nfultz) [inline R/diagnosis_helper_functions.R]: Rewrite using Reduce

File: R/quick_design.R [hunk 16]
```diff
@@ -36,26 +36,39 @@
...
   else
     cbind.data.frame(..., stringsAsFactors=FALSE)
-
   k <- nrow(template_args_matrix)
-  if(k == 1) return( template(...) )
 
-  by(template_args_matrix, 1:k, do.call, what=template, simplify = FALSE)
+  if(k == 1) {
+    design_args <- list(...)
+    return(do.call(template, design_args))
```
Comment (nfultz) [inline R/quick_design.R]: If doing do.call anyway, may as well fall through instead of special casing it.

File: R/quick_design.R [hunk 17]
```diff
@@ -36,26 +36,39 @@
...
+  } else {
+    out <- by(template_args_matrix, 1:k, do.call, what=template, simplify = FALSE)
+  }
+
+  # Add attribute
+  for(j in 1:k) {attr(out[[j]], "parameters") <-
+    setNames(template_args_matrix[j,], names(template_args_matrix) )}
+
+  names(out) <- paste0(name_stub, 1:length(out))
```
Comment (nfultz) [inline R/quick_design.R]: 1:k

File: R/utilities.R [hunk 18]
```diff
@@ -139,3 +139,28 @@ rbind_disjoint <- function(list_of_df, infill=NA) {
 `%i%` <- intersect
 
 `%||%` <- function(e1, e2) if(is.null(e1)) e2 else e1
+
+add_parens <- function(x, digits = 3) {
+  x <- as.numeric(x)
```
Comment (nfultz) [inline R/utilities.R]: no need for as.numeric here, format_nums handles it. also use implicit return.

File: R/utilities.R [hunk 19]
```diff
@@ -139,3 +139,28 @@ rbind_disjoint <- function(list_of_df, infill=NA) {
...
+  return_vec <- sprintf("(%s)", format_num(x, digits))
+  return(return_vec)
+}
+
+format_num <- function(x, digits = 3) {
+  x <- as.numeric(x)
+  return(paste0(sprintf(paste0("%.", digits, "f"), x)))
```
Comment (nfultz) [inline R/utilities.R]: gratuituous paste0

**Review by nfultz — COMMENTED** [review]
Overall seems pretty good. `diagnose_design` should be further split up, probably as a generic, now that simulation is decoupled from it. 

Also please check your editor settings, many variable assignmentss are being split across multiple lines incorrectly. We are trying to match the internal R style.

**Comment by graemeblair** [comment]
made many small changes throughout; overall looks good

addressed all of Neal's comments except the move to s3 -- did it, but then the s3 methods didn't send the names of the objects on to rlang properly so had to revert. let's do in future.

**Review by acoppock — APPROVED** [review]
lgtm

**Pull request merged**